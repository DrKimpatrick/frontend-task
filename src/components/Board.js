import React, {Component} from 'react';
import Cell from './cell';
import sprite from '../images/green.jpg';

export default class Board extends Component {
  state = {
      count:1,
        zero: '',
        possibleTopIdx: '',
        possiblRightIdx: '',
        possiblBottomIdx: '',
        possibleLeftIdx: ''
    };
    componentWillMount() {
        this.findClickables(this.props.board, this.props.size);
    }
    componentWillReceiveProps(nextProps) {
        this.findClickables(nextProps.board, nextProps.size);
    }
    shouldComponentUpdate(nextProps) {
        const curr = this
            .props
            .board
            .join('');
        const next = nextProps
            .board
            .join('');
        return curr !== next;
    }

    findClickables = (board, size) => {
        const zeroIndex = board.indexOf(0);
        const zeroCoordinate = this.getCoordFromIndex(zeroIndex, size);
        const possibleTopIdx = zeroCoordinate.row > 0
            ? this.getIndexFromCoord(zeroCoordinate.row - 1, zeroCoordinate.column, size)
            : null;
        const possiblRightIdx = zeroCoordinate.column < size
            ? this.getIndexFromCoord(zeroCoordinate.row, zeroCoordinate.column + 1, size)
            : null;
        const possiblBottomIdx = zeroCoordinate.row < size
            ? this.getIndexFromCoord(zeroCoordinate.row + 1, zeroCoordinate.column, size)
            : null;
        const possibleLeftIdx = zeroCoordinate.column > 0
            ? this.getIndexFromCoord(zeroCoordinate.row, zeroCoordinate.column - 1, size)
            : null;

        this.setState({zero: zeroIndex, possibleTopIdx: possibleTopIdx, possiblRightIdx: possiblRightIdx, possiblBottomIdx: possiblBottomIdx, possibleLeftIdx: possibleLeftIdx});
    }
    getCoordFromIndex = (idx, size) => {
        return {
            row: Math.floor(idx / size) + 1,
            column: (idx % size) + 1
        };
    }
    getIndexFromCoord = (row, col, size) => {
        return (size * (row - 1)) + col - 1;
    }
    cellClickHandler = (index) => {
        if (index === this.state.possibleTopIdx || index === this.state.possiblRightIdx || index === this.state.possiblBottomIdx || index === this.state.possibleLeftIdx) 
            this.nextBoard(index);
        }
  nextBoard = (index) => {
    this.setState({ count: this.state.count + 1 })

        const indexx = this.props.sprite.indexOf(index)
        if (indexx > -1) {
          this.props.sprite.splice(indexx, 1)
        }
        if (this.props.sprite.length === 0) {
          alert(`Game Over. Total moves to save bae: ${this.state.count}`)
          window.location.href = '/'
        }

        // console.log(this.props.sprite)
        const board = this
            .props
            .board
            .slice();
        const temp = board[index];
        board[index] = board[this.state.zero];
        board[this.state.zero] = temp;
        this
            .props
            .updateBoard(board);
    }
    render() {
        const squares = this
            .props
            .board
            .map((val, index) => {

                if ((index + 1) % this.props.size === 0) {
                    if (this.props.sprite.includes(val)) {
                        return (
                            <span key={'i' + index}>
                              <img
                        key={index}   
                              alt='sprite'
                      style={{width:60,height:60}}
                       src={sprite}
                        onClick={() => this.cellClickHandler(index)} />
                                <br/>
                            </span>
                        )
                    } else {
                        return (
                            <span key={'i' + index}>
                                {< Cell key = {
                                    index
                                }
                                value = {
                                    val
                                }
                                sprite = {
                                    this.props.sprite
                                }
                                clickHandler = {
                                    () => this.cellClickHandler(index)
                                } />}
                                <br/>
                            </span>
                        );
                    }

                }
                if (this.props.sprite.includes(val)) {
                    return (<img
                      key={index}
                      alt='sprite'
                      style={{width:60,height:60}}
                      src={sprite}
                      onClick={() => this.cellClickHandler(index)} />
                      )
                } else {
                    return <Cell
                        key={index}
                        value={val}
                        clickHandler={() => this.cellClickHandler(index)}/>;
                }

            });
        return (
            <div className='board'>
                {squares}
            </div>
        );
    }
}
