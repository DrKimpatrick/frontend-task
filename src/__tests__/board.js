import React from 'react';
import { shallow } from 'enzyme';
import Board from '../components/Board';

describe('Board component', () => {
  it('Should render Board component', () => {
    const component = shallow(<Board size={2} board={[3,3]} updateBoard={()=> {}} sprite={[3]} />);
  
    expect(component).toMatchSnapshot();
  });
  
});