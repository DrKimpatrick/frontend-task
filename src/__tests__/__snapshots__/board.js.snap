// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board component Should render Board component 1`] = `
<div
  className="board"
>
  <img
    alt="sprite"
    key="0"
    onClick={[Function]}
    src="green.jpg"
    style={
      Object {
        "height": 60,
        "width": 60,
      }
    }
  />
  <span
    key="i1"
  >
    <img
      alt="sprite"
      key="1"
      onClick={[Function]}
      src="green.jpg"
      style={
        Object {
          "height": 60,
          "width": 60,
        }
      }
    />
    <br />
  </span>
</div>
`;
